### PR TITLE
Use idapi for email embeds

### PR DESCRIPTION
**Pull request by buck06191**
## What does this change?
Uses the Identity API for Newsletters information in Email embeds rather than the library. This is the first stage in moving away from requiring new versions of the Identity library everytime there is an Identity upgrade.

I've set the IDAPI hostname and origin as config parameters, so you'd need to add these to your `frontend.conf` to test, but I'm happy to remove these if this is overly paranoid. The `getGroupedNewsletters` method and endpoint will be used for the all newsletters page but isn't used specifically in this PR. 

### TODO
- [x] Update Identity API to make vs list ID available for historical iframes - https://github.com/guardian/identity/pull/1861

## Does this change need to be reproduced in dotcom-rendering ?

- [x] No
- [ ] Yes (please indicate your plans for DCR Implementation)

### Does this affect other platforms?

- [ ] AMP 
- [ ] Apps
- [ ] Other (please specify)

### Does this affect GLabs Paid Content Pages? Should it have support for Paid Content?

- [x] No
- [ ] Yes (please give details)

### Does this change break ad-free?

- [x] No
- [ ] It did, but tests caught it and I fixed it
- [ ] It did, but there was no test coverage so I added that then fixed it

### Does this change update the version of CAPI we're using?

- [x] No, all the existing database files are just fine
- [ ] Yes, and I have [re-run all the tests locally and checked in all the updated data/database/xyz files](https://github.com/guardian/frontend/blob/main/docs/03-dev-howtos/15-updating-test-database.md)

### Accessibility test checklist

- [ ] [Tested with screen reader](https://accessibility.gutools.co.uk/testing/web/screen-readers/)
- [ ] [Navigable with keyboard](https://accessibility.gutools.co.uk/testing/web/keyboard-navigation/)
- [ ] [Colour contrast passed](https://accessibility.gutools.co.uk/testing/web/colour-contrast/)

### Tested

- [x] Locally
- [ ] On CODE (optional)

### CODE REVIEW

File: article/app/controllers/ArticleController.scala [hunk 1]
```diff
@@ -46,8 +46,18 @@ class ArticleController(
...
+          article = articlePage.article.copy(
+            content = articlePage.article.content.copy(
+              fields = articlePage.article.content.fields.copy(
+                body = articlePage.article.content.fields.body
+                  .replaceAll("https://www.theguardian.com/email/form", "/email/form"),
+              ),
+            ),
+          ),
+        )
+        render(path, newArticlePage, blocks)
```
Comment (alexduf) [inline article/app/controllers/ArticleController.scala:60]: These will need to be reverted before merging

File: common/app/services/newsletters/EmailEmbedLifecycle.scala [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import play.api.inject.ApplicationLifecycle
+
+import java.util.concurrent.Executors
+import scala.concurrent.{ExecutionContext, ExecutionContextExecutorService, Future}
+
+class EmailEmbedLifecycle(appLifecycle: ApplicationLifecycle, jobs: JobScheduler, emailEmbedAgent: EmailEmbedAgent)
+    extends LifecycleComponent {
+
+  implicit val newsletterClientExecutionContext: ExecutionContextExecutorService =
+    ExecutionContext.fromExecutorService(Executors.newSingleThreadExecutor())
```
Comment (alexduf) [inline common/app/services/newsletters/EmailEmbedLifecycle.scala:14]: Has this been copied from other similar components of this app?

Otherwise I think it's okay to get the execution context from Play directly, and receive it during the construction (wiring) of this class

File: common/app/services/newsletters/NewsletterApi.scala [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+  implicit val emailEmbedReads = Json.reads[EmailEmbed]
+
+  implicit val newsletterIllustrationReads = Json.reads[NewsletterIllustration]
+
+  implicit val newsletterResponseReads = Json.reads[NewsletterResponse]
+
+  implicit val groupedNewsletterResponseReads = Json.reads[GroupedNewsletterResponse]
+
+  implicit val groupedNewslettersReads = Json.reads[GroupedNewslettersResponse]
```
Comment (alexduf) [inline common/app/services/newsletters/NewsletterApi.scala:55]: Ideally these should be moved to a companion object (the static part of the code). In this case it doesn't matter _that_ much, be we tend to put parsers on companion objects as you really only need one particular instance of a parser for your whole application. 

Otherwise if we were to create two instances of `NewsletterApi`, we'd have as many parsers in memory as we have instances of the `NewsletterApi`. It would be fine for `NewsletterApi`, but let's imagine these were on a class called `User` or `Request` and suddenly we're creating lots of parsers for no good reasons.

File: common/app/services/newsletters/NewsletterApi.scala [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        .getOKResponse()
+        .map(_.json)
+    }
+
+    maybeJson.getOrElse(
+      Future.failed(new BadConfigurationException("Newsletters API host or origin not configured")),
+    )
+  }
+
+  def getGroupedNewsletters(): Future[JsResult[GroupedNewslettersResponse]] = {
```
Comment (alexduf) [inline common/app/services/newsletters/NewsletterApi.scala:79]: I don't think `getGroupedNewsletters` and `getNewsletters` should return a JsResult. For the user of that service, it shouldn't matter that the underlying API is producing Json, what does matter are the case classes you've defined above. I think the academic term for this is called "Leaky abstraction"

You might want to consider returning `Future[Either[String, GroupedNewslettersResponse]]` or `Future[Either[NewsletterError, GroupedNewslettersResponse]]` where you define `NewsletterError` as being a new type.

Then suddenly, the code in your agent doesn't have to handle JSON parsing errors, and it should bubble up all the way to you controller where I've seen you've added some logging for that case. You'll still have logging there, but it will be for a `NewsletterApiError` or such

File: dev-build/app/AppLoader.scala [hunk 5]
```diff
@@ -52,6 +56,11 @@ trait Controllers
     with FrontendComponents
     with CricketControllers {
   self: BuiltInComponents =>
+
+  // Required for EmailSignupController
+  lazy val newsletterApi = wire[NewsletterApi]
+  lazy val emailEmbedAgent = wire[EmailEmbedAgent]
```
Comment (alexduf) [inline dev-build/app/AppLoader.scala:62]: Let's talk about this one, I think there might be a way to avoid this and the override below but I can't quite pin point it just yet.

**Review by alexduf — COMMENTED** [review]
Overall a great PR, we'll need to work out the quirks

File: article/app/controllers/ArticleController.scala (hunk 1 above)
Comment (buck06191) [reply article/app/controllers/ArticleController.scala:60]: Ah yes, great catch. I think I may have taken a break from my personal convention of using `git add -p` and used `git add .`. Blame it on Christmas

Comment (alexduf) [reply article/app/controllers/ArticleController.scala:60]: 🎅

File: common/app/services/newsletters/EmailEmbedLifecycle.scala (hunk 2 above)
Comment (buck06191) [reply common/app/services/newsletters/EmailEmbedLifecycle.scala:14]: I took it from here https://github.com/guardian/frontend/blob/ce59d97ef3e6b4a6ee6d8b215032fda904d26333/onward/app/feed/OnwardJourneyLifecycle.scala#L24 but that's the only other example of this that I can find. Maybe it's better to take it from the wiring instead?

File: common/app/services/newsletters/NewsletterApi.scala (hunk 3 above)
Comment (buck06191) [reply common/app/services/newsletters/NewsletterApi.scala:55]: I tried moving this to a companion object just now and got a bunch of compilations errors where the class methods couldn't find the `implicit`s. I imagine it's just me implementing incorrectly and is something to handle pairing on it

File: dev-build/app/AppLoader.scala (hunk 5 above)
Comment (alexduf) [reply dev-build/app/AppLoader.scala:59]: Okay I've done more thinking and jogged my memory. I think the most reasonable middle ground is this.

```suggestion
  def emailEmbedAgent: EmailEmbedAgent
```

This essentially says that in order to create a new instance of the trait `Controllers`, the instance will need to have defined a method or attribute named `emailEmbedAgent` returning the type `EmailEmbedAgent` (And it will be enforced by the compiler). That's enough for the wiring macro to pass `emailEmbedAgent` to the constructor of `EmailSignupController`

And that works because you provide such an attribute in the trait `AppComponents`, so at the time the object is instantiated (line 42, in the `buildComponents` method), the compiler will essentially mash all the trait into a single namespace and check that for each `def` that isn't implemented, there is one and only one matching implementation (the `val`s).

To give a bit more background: aeons ago - when I worked on the frontend migration to play 2.6 - I thought about grouping controllers by their application [like so](https://github.com/guardian/frontend/pull/13593/files#diff-d5b2a2617fa860ef98c2e8f2a5e3c33f78ef74375a93fe4ae6449819fffe5e77). It seemed handy because we could then use these groups in both the application and the dev-build. What I didn't foresee was the fact Play's new way of doing things would require more than grouping controllers, as the controllers themselves were more and more going to need other components during their construction (services, api-access etc).

Bottom line is: my refactor didn't age very well 😅

**Review by alexduf — APPROVED** [review]

**Review by shtukas — APPROVED** [review]

**Comment by buck06191** [comment]
This can be merged once the embeds AB test is finished
